### ISSUE TITLE
平均发放速率

### ISSUE THREAD

**Issue by chenjiefighting** [fangwei123456/spikingjelly#285]
你好，我想问一下，随着网络深度的增加，脉冲平均发放频率应该成什么样的趋势？

**Comment by fangwei123456** [comment]
不做任何矫正的话，容易出现趋向于0

**Comment by yizexiang123456** [comment]
> 不做任何矫正的话，容易出现趋向于0

请问一下，有哪些矫正的方法呢？

**Comment by fangwei123456** [comment]
> > 不做任何矫正的话，容易出现趋向于0
> 
> 请问一下，有哪些矫正的方法呢？

参考最近几年顶会里面涉及SNN的normalization的文章

**Comment by chenjiefighting** [comment]
随着网络深度的增加，脉冲平均发放频率趋于零，请问一下，这和梯度消失是一回事吗？这两者之间有什么联系？

**Comment by fangwei123456** [comment]
有一定的关系，如果梯度正常，那网络应该能正常训练；能正常训练的网络就不会出现深层一直输出0的情况

**Comment by chenjiefighting** [comment]
我搭建了一个脉冲神经网络，训练时统计了每一层的脉冲平均发放频率，发现从5层开始脉冲平均发放频率为0了，最终结果也都分为一类，按照你的说法这种情况应该不是梯度消失，可以帮忙分析一下吗？

**Comment by yizx6** [comment]
可以在cov后面加bn，[参考这个](https://github.com/fangwei123456/Parametric-Leaky-Integrate-and-Fire-Spiking-Neuron)。

**Comment by fangwei123456** [comment]
梯度不消失只是收敛的必要条件，所以有梯度但不收敛也是正常的

**Comment by chenjiefighting** [comment]
你好，我还有一个问题，目前我使用公式avg_firing_rate = spikes/(neuronT*batchsize)得到网络每一层的平均脉冲发放率，请问如何估算网络的运算量？

**Comment by fangwei123456** [comment]
运算量用flop相关的计算库，例如
https://github.com/Lyken17/pytorch-OpCounter

**Comment by chenjiefighting** [comment]
感谢回复，如果是从理论分析ANN的计算开销主要是乘法操作和加法操作，而SNN的计算开销主要是加法操作，请问SNN的可以这样进行估算吗？
![image](https://user-images.githubusercontent.com/76989858/217524672-2fa949a7-66ce-4301-af41-206400d1077d.png)

**Comment by fangwei123456** [comment]
> 感谢回复，如果是从理论分析ANN的计算开销主要是乘法操作和加法操作，而SNN的计算开销主要是加法操作，请问SNN的可以这样进行估算吗

原理是这样的

**Comment by chenjiefighting** [comment]
可能是我的表述有问题，其实我想问的是如何用平均脉冲发放率估算SNN的加法操作次数，是在公式1.10的基础上乘平均脉冲发放率吗？

**Comment by fangwei123456** [comment]
这个我不确定，我并没有检查这个公式的正确性。

**Comment by fangwei123456** [comment]
有一个本组师兄做的开源工具，可以试试

https://github.com/iCGY96/syops-counter

**Comment by chenjiefighting** [comment]
好嘞，太感谢了

**Comment by AllenYolk** [comment]
深度 SNN 中膜平均发放率随深度增加而衰减是常见现象，主要原因：

1. **激活值分布漂移**。前层 spike 频率较低时，后层累积电流不足，膜电位长时间低于阈值，firing rate 自然趋于 0。
2. **梯度消失 / surrogate gradient 饱和**。Surrogate 在膜电位远离阈值时梯度接近 0，反向信号无法穿透。
3. **膜电位衰减过强**。LIF 的 τ 偏小也会让 v 衰减得过快。

常用矫正手段：

- **BatchNorm / tdBN / BNTT**：在 conv/fc 后插入 BN，将膜电位分布重置到均值为 0 方差为 1 的范围，对 firing rate 的稳定效果非常显著。
- **阈值 / τ 调整**：减小 v_threshold 或增大 τ，让膜电位更容易跨过阈值。
- **shortcut / SEW block**：残差连接是缓解深度 SNN 退化的有效手段（参考 SEW-ResNet）。
- **TOSS / PTQ**：训练后阈值 / 权重均衡。

参考：
- TDNN / tdBN：https://arxiv.org/abs/2011.05280
- SEW-ResNet：https://arxiv.org/abs/2102.04059

估算网络运算量可以用 `pytorch-OpCounter`（理论 FLOPs）或 `iCGY96/syops-counter`（结合 spike rate 得到 spike-driven 的 AC operations）。
